Ledgerstate API docs (#1346)
iotaledger/goshimmer: docs/apis/ledgerstate.md

* [client lib - `GetTransactionMetadata()](#client-lib---gettransactionmetadata)
* [client lib - `GetTransactionInclusionState()](#client-lib---gettransactioninclusionstate)
* [client lib - `GetTransactionConsensusMetadata()](#client-lib---gettransactionconsensusmetadata)
* [client lib - `GetTransactionAttachments()](#client-lib---gettransactionattachments)
* [client lib - `PostTransaction()](#client-lib---posttransaction)
* [client lib - `PostAddressUnspentOutputs()](#client-lib---postaddressunspentoutputs)

## `/ledgerstate/addresses/:address`
Get an address details for a given base58 encoded address ID, such as output types and balances. For the client library API call balances will not be directly available as values because they are stored as a raw message. Balance cann be read after retrieving `ledgerstate.Output` instance, as presented in the examples.

### Parameters
| **Parameter**            | `addressID`      |
|--------------------------|----------------|
| **Required or Optional** | required       |
| **Description**          | The address ID encoded in base58. |
| **Type**                 | string         |

### Examples

#### cURL

```shell

> curl http://localhost:8080/ledgerstate/addresses/:address \

Reviewer: I would keep :address, but state in the line thereafter
where `:address` is the address, e.g. 6PQqFcwarCVbEMxWFeAqj7YswK842dMtf84qGyKqVH7s1kK 
because it is useful to highlight that the string that goes there is the address. See also my comment on `:branchID`

-X GET \
-H 'Content-Type: application/json'
```
#### client lib - `GetAddressOutputs()`
```Go
resp, err := goshimAPI.GetAddressOutputs("6PQqFcwarCVbEMxWFeAqj7YswK842dMtf84qGyKqVH7s1kK")
	if err != nil {
		// return error
	}
	fmt.Println("output address: ", resp.Address)
	
	for _, output := range resp.Outputs {
		fmt.Println("outputID: ", output.OutputID)
		fmt.Println("output type: ", output.Type)
		// get output instance

> 		out, err = output.ToLedgerstateOutput()

Reviewer 2: Adjust indent.

| `base58`| string   | The address encoded with base58.          |

`Output`

|Field | Type | Description|
|:-----|:------|:------|
| `outputID`  | OutputID | The identifier of an output.   |
| `outputType`   | string | The full ID of a node.     |
| `output` | string | An output raw message containing balances and corresponding addresses. |

`OutputID`

|Field | Type | Description|
|:-----|:------|:------|
| `base58`  | string | The output identifier encoded with base58.    |
| `transactionID`   | string | The transaction identifier encoded with base58.     |
| `outputIndex`   | int | The index of an output.     |

## `/ledgerstate/addresses/:address/unspentOutputs`
Gets list of all unspent outputs for the address based on a given base58 encoded address ID.

### Parameters

| **Parameter**            | `addressID`      |
|--------------------------|----------------|
| **Required or Optional** | required       |
| **Description**          | The address ID encoded in base58. |
| **Type**                 | string         |
### Examples

#### cURL

```shell
curl http://localhost:8080/ledgerstate/addresses/:address/unspentOutputs \
-X GET \
-H 'Content-Type: application/json'
```

> #### client lib - `GetAddressUnspentOutputs()`

Reviewer: This is hard to read in the rendered version. We should change this from 
#### client lib - `GetAddressUnspentOutputs()`
to a bullet point
* client lib - `GetAddressUnspentOutputs()`
for all examples (not only in this document)

Author: Don't we loose the links to the methods in the table of content?

for _, output := range resp.Outputs {
    fmt.Println("outputID: ", output.OutputID)
    fmt.Println("output type: ", output.Type)
    // get output instance
    out, err = output.ToLedgerstateOutput()
}
```
### Response examples
```json
{
    "address": {
        "type": "AddressTypeED25519",
        "base58": "18LhfKUkWt4M9YR6Q3au4LT8wWCERwzHaqn153K78Eixp"
    },
    "outputs": [
        {
            "outputID": {
                "base58": "gdFXAjwsm5kDeGdcZsJAShJLeunZmaKEMmfHSdoX34ZeSs",
                "transactionID": "32yHjeZpghKNkybd2iHjXj7NsUdR63StbJcBioPGAut3",
                "outputIndex": 0
            },
            "type": "SigLockedColoredOutputType",
            "output": {
                "balances": {
                    "11111111111111111111111111111111": 1000000
                },
                "address": "18LhfKUkWt4M9YR6Q3au4LT8wWCERwzHaqn153K78Eixp"
            }
        }
    ]
}
```

### Results
|Return field | Type | Description|
|:-----|:------|:------|
| `address`  | Address | The address corresponding to provided unspent outputID.   |
| `outputs`   | Output | List of transactions' unspent outputs.     |

> `Address`

Author: I added  "#### Type" here and in mana docs, it looks nice

| `branchID`  | string | The branch identifier encoded with base58.   |
| `type`        | string | The type of the branch.  |

## `/ledgerstate/branches/:branchID/conflicts`
Get all conflicts for a given branch ID, their outputs and conflicting branches.

### Parameters

| **Parameter**            | `branchID`      |
|--------------------------|----------------|
| **Required or Optional** | required       |
| **Description**          | The conflicting branch ID encoded in base58. |
| **Type**                 | string         |

### Examples

#### cURL

```shell
curl http://localhost:8080/ledgerstate/branches/:branchID/conflicts \
-X GET \
-H 'Content-Type: application/json'
```
#### client lib - `GetBranchConflicts()`
```Go
	resp, err := goshimAPI.GetBranchConflicts("2e2EU6fhxRhrXVnYQ6US4zmUkE5YJip25ecafn8gZeoZ")
	if err != nil {
		// return error
	}
	fmt.Printf("All conflicts for branch %s:\n", resp.BranchID)
    // iterate over all conflicts
	for _, branch := range resp.Conflicts {
		fmt.Println("output ID: ", branch.OutputID.Base58)
		fmt.Println("conflicting transaction ID: ", branch.OutputID.TransactionID)
		fmt.Printf("related branches: %v\n", branch.BranchIDs)
	}
```

> ### Response examples

Reviewer 2: Adjust indent.

fmt.Printf("finalized: %v, solid: %v\n", resp.Finalized, resp.Solid)
fmt.Println("solidification time: ",  time.Unix(resp.SolidificationTime, 0))
```
### Response examples
```json
{
    "outputID": {
        "base58": "41GvDSQnd12e4nWnd2WzmdLmffruXqsE46jgeUbnB8s1QnK",
        "transactionID": "9wr21zza46Y5QonKEHNQ6x8puA7Rbq5LAbsQZJCK1g1g",
        "outputIndex": 0
    },
    "branchID": "4uQeVj5tqViQh7yWWGStvkEG1Zmhx6uasJtWCJziofM",
    "solid": true,
    "solidificationTime": 1621889327,
    "consumerCount": 2,
    "firstConsumer": "HuYUAwCeexmBePNXx5rNeJX1zUvUdUUs5LvmRmWe7HCV",
    "finalized": true
}
```

### Results

|Return field | Type | Description|
|:-----|:------|:------|
| `outputID`            | OutputID  | The output identifier encoded with base58.   |
| `branchID`            | string    | The identifier of the branch encoded with base58. |
| `solid`               | bool      | The boolean indicator if the message is solid. |
| `solidificationTime`  | int64     | The time of solidification of a message. |
| `consumerCount`       | int       | The number of consumers. |
| `firstConsumer`       | string    | The first consumer of the output. |

> | `finalized`           | bool      | The boolean indicator if the transaction . |

Reviewer 2: ```suggestion
| `finalized`           | bool      | The boolean indicator if the transaction finalized. |
```

            "output": {
                "balances": {
                    "11111111111111111111111111111111": 1000000
                },
                "address": "1HrUn1jWAjrMU58LLdFhfnWBwUKVdWjP5ojp7oCL9mVWs"
            }
        }
    ],
    "unlockBlocks": [
        {
            "type": "SignatureUnlockBlockType",
            "publicKey": "12vNcfgRHLSsobeqZFrjFRcVAmFQbDVniguPnEoxmkbG",
            "signature": "4isq3qzhY4MwbSeYM2NgRn5noWAyh5rqD12ruiTQ7P89TfXNecwHZ5nbpDc4UB7md1bkfM1xYtSh18FwLqK8HAC6"
        }
    ],
    "dataPayload": ""
}
```

### Results
|Return field | Type | Description|
|:-----|:------|:------|
| `version`         | uint8  | The version of the transaction essence.   |
| `timestamp`       | int64    | The issuing time of the transaction. |
| `accessPledgeID`  | string      | The boolean indicator if the message is solid. |
| `consensusPledgeID`| string     | The time of solidification of a message. |
| `inputs`          | []Input       | The number of consumers. |
| `outputs`         | []Output    | The firs consumer of athe output. |
| `unlockBlocks`    | []UnlockBlock      | The boolean indicator if the transaction . |

> | `dataPayload` | []byte      | The boolean indicator if the transaction . |

Reviewer 2: The descriptions here are incorrect.

`Input `
|Field | Type | Description|
|:-----|:------|:------|
| `Type`  | string | The type of input.   |
| `ReferencedOutputID`   | string | The output ID that is used as an input for the transaction.     |

> | `output` | Output | An output that is used as an input for the transaction |

Reviewer 2: There's no `Output` in the response above.
And add a section for `ReferencedOutputID`.

| **Description**          | The transaction ID encoded in base58. |
| **Type**                 | string         |

### Examples

#### cURL

```shell
curl http://localhost:8080/ledgerstate/transactions/:transactionID/metadata \
-X GET \
-H 'Content-Type: application/json'
```
#### client lib - `GetTransactionMetadata()`
```Go
resp, err := goshimAPI.GetTransactionMetadata("41GvDSQnd12e4nWnd2WzmdLmffruXqsE46jgeUbnB8s1QnK")
if err != nil {
    // return error
}
fmt.Println("transactionID:", resp.TransactionID)
fmt.Println("branchID:", resp.BranchID)
fmt.Printf("branch lazy booked: %v, solid: %v, finalized: %v\n", resp.LazyBooked, resp.Solid, resp.Finalized)
fmt.Println("solidification time:",  time.Unix(resp.SolidificationTime, 0))
```
### Response examples
```json
{
    "transactionID": "HuYUAwCeexmBePNXx5rNeJX1zUvUdUUs5LvmRmWe7HCV",
    "branchID": "HuYUAwCeexmBePNXx5rNeJX1zUvUdUUs5LvmRmWe7HCV",
    "solid": true,
    "solidificationTime": 1621889358,
    "finalized": true,
    "lazyBooked": false
}
```
### Results
|Return field | Type | Description|
|:-----|:------|:------|
| `transactionID`         | string  | The transaction identifier encoded with base58.    |

> | `branchID`       | string    | The issuing time of the transaction. |

Reviewer 2: ```suggestion
| `branchID`       | string    | The branch identifier of the transaction. |
```

if err != nil {
    // return err
}
fmt.Printf("Consensus metadata for transaction %s:\n", resp.TransactionID)
fmt.Println("fcob opinion, is liked:", resp.Liked)
fmt.Println("level of knowledge:", resp.LoK)
fmt.Println("fcob timestamp:", time.Unix(resp.Timestamp, 0))
fmt.Println("fcob time 1:", time.Unix(resp.FCOBTime1, 0))
fmt.Println("fcob time 2:", time.Unix(resp.FCOBTime2, 0))
```

### Response examples
```json
{
    "transactionID": "HuYUAwCeexmBePNXx5rNeJX1zUvUdUUs5LvmRmWe7HCV",
    "timestamp": 1621889358,
    "liked": true,
    "lok": "LevelOfKnowledge(Three)",
    "fcobTime1": 1621889360,
    "fcobTime2": 1621889362
}
```

### Results
|Return field | Type | Description|
|:-----|:------|:------|
| `transactionID`         | string  | The transaction identifier encoded with base58.  |
| `timestamp`       | uint64    | The timestamp of the arrival. |
| `lok`          | string      | The level of knowledge of the transaction - the degree of certainty of the associated opinion.|
| `liked`  | bool      | The fcob opinion for the transaction. |
| `fcobTime1`         | bool    | The fcob opinion's execution time.|

> | `fcobTime2`         | bool    | The fcob opinion's locally finalized execution time.|

Reviewer 2: Type: uint64.

### Parameters
| **Parameter**            | `transactionID`      |
|--------------------------|----------------|
| **Required or Optional** | required       |
| **Description**          | The transaction ID encoded in base58. |
| **Type**                 | string         |

### Examples

#### cURL

```shell
curl http://localhost:8080/ledgerstate/transactions/:transactionID/attachments \
-X GET \
-H 'Content-Type: application/json'
```
#### client lib - `GetTransactionAttachments()`
```Go
resp, err := goshimAPI.GetTransactionAttachments("DNSN8GaCeep6CVuUV6KXAabXkL3bv4PUP4NkTNKoZMqS")
if err != nil {
    // return error
}
fmt.Printf("Messages IDs containing transaction %s:\n", resp.TransactionID)
for _, msgID := range resp.MessageIDs {
    fmt.Println(msgID)
}
```
### Response examples
```json
{
    "transactionID": "HuYUAwCeexmBePNXx5rNeJX1zUvUdUUs5LvmRmWe7HCV",
    "messageIDs": [
        "J1FQdMcticXiiuKMbjobq4zrYGHagk2mtTzkVwbqPgSq"
    ]
}
```

### Results
|Return field | Type | Description|
|:-----|:------|:------|
| `transactionID`   | string  | The transaction identifier encoded with base58.  |

> | `messageIDs`       | []string    | The messages IDs that contains the requested transaction payload. |

Reviewer 2: ```suggestion
| `messageIDs`       | []string    | The messages IDs that contains the requested transaction. |
```

                            "transactionID": "BqzgVk4yY9PDZuDro2mvT36U52ZYbJDfM41Xng3yWoQK",
                            "outputIndex": 0
                        },
                        "type": "SigLockedColoredOutputType",
                        "output": {
                            "balances": {
                                "11111111111111111111111111111111": 1000000
                            },
                            "address": "1Z4t5KEKU65fbeQCbNdztYTB1B4Cdxys1XRzTFrmvAf3"
                        }
                    },
                    "inclusionState": {
                        "confirmed": true,
                        "rejected": false,
                        "conflicting": false
                    },
                    "metadata": {
                        "timestamp": "2021-05-25T15:47:04.50470213+02:00"
                    }
                }
            ]
        }
    ]
}
```
### Results

|Return field | Type | Description|
|:-----|:------|:------|
| `unspentOutputs`   | WalletOutputsOnAddress  | Unspent outputs representation for wallet.  |

`WalletOutputsOnAddress`
|Return field | Type | Description|
|:-----|:------|:------|

> | `Address`   | Address  | Unspent outputs representation for wallet.  |

Reviewer 2: Incorrect description here.

| `Outputs`   | []WalletOutput  | Unspent outputs representation for wallet.  |

`Address`

|Field | Type | Description|
|:-----|:------|:------|
| `type`  | string | The type of an address.   |
| `base58`| string   | The address encoded with base58.          |

`WalletOutput`
|Field | Type | Description|
|:-----|:------|:------|

> | `output`  | Output | The type of address.   |

Reviewer 2: ```suggestion
| `output`  | Output | The unspent output.   |
```

| `inclusionState`| InclusionState   | The inclusion state of the transaction containing the output.  |
| `metadata`| WalletOutputMetadata   | The metadata of the output for the wallet lib.  |

`Output`

|Field | Type | Description|
|:-----|:------|:------|
| `outputID`  | OutputID | The identifier of an output.   |

> | `outputType`   | string | The full ID of a node.     |

Reviewer 2: ```suggestion
| `outputType`   | string | The type of the output.     |
```
